### PR TITLE
Enable authorizedDatabases & authorizedCollections for fetching metadata in MongoDB

### PR DESCRIPTION
**Pull request by huberty89**
## Description
Enable authorizedDatabases & authorizedCollections for fetching metadata in MongoDB

## Fixes
https://github.com/trinodb/trino/issues/1398

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# MongoDB
* Fix failure when an unauthorized exception happens while listing schemas or tables. ({issue}`1398`)
```

### CODE REVIEW

**Comment by huberty89** [comment]
I need to update the docs

**Comment by huberty89** [comment]
Looks like `TestMetadataPrivilegesRequired` test added by me is flaky - Docker container with MongoDB cannot starts successfully.
I will take a look on Monday.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoClientConfig.java [hunk 1]
```diff
@@ -44,6 +44,7 @@
     private Optional<String> connectionUrl = Optional.empty();
     private List<ServerAddress> seeds = ImmutableList.of();
     private List<MongoCredential> credentials = ImmutableList.of();
+    private boolean metadataPrivilegesRequired = true;
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoClientConfig.java:47]: Could you share why we want to introduce a config property? Other connectors hide unauthorized databases and tables. Always enabling `authorizedDatabasesOnly` & `authorizedCollections` properties looks fine to me.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 2]
```diff
@@ -823,34 +834,62 @@ private String toRemoteSchemaName(String schemaName)
...
             }
         }
         return schemaName;
     }
 
+    private Iterable<String> listDatabaseNames()
+    {
+        OperationExecutor operationExecutor = ((MongoClientImpl) client).getOperationExecutor();
+        CodecRegistry codecRegistry = ((MongoClientImpl) client).getSettings().getCodecRegistry();
+        boolean retryReads = ((MongoClientImpl) client).getSettings().getRetryReads();
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:849]: nit: I would extract a variable:
```java
MongoClientImpl mongoClient = (MongoClientImpl) client;
```

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java [hunk 3]
```diff
@@ -29,10 +31,16 @@ public MongoServer()
     }
 
     public MongoServer(String mongoVersion)
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java:33]: The usage of this constructor is only 2 places. I would change this constructor instead of adding a new constructor.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java [hunk 4]
```diff
@@ -29,10 +31,16 @@ public MongoServer()
...
+    {
+        this(mongoVersion, ImmutableMap.of());
+    }
+
+    public MongoServer(String mongoVersion, Map<String, String> env)
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java:38]: nit
```suggestion
    public MongoServer(String mongoVersion, Map<String, String> environmentVariables)
```

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+    public void testTablesVisibility()
+    {
+        assertThat(queryRunner.execute("SHOW TABLES FROM mongodb." + TEST_DATABASE)
+                .getOnlyColumnAsSet())
+                .containsExactly(TEST_COLLECTION.toLowerCase(Locale.ENGLISH));
+    }
+
+    private static MongoServer setupMongoServer()
+    {
+        MongoServer server = new MongoServer("4.0.6", ImmutableMap.of(
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java:115]: MongoDB 4.0 already reached EOL on Apr 30. Let's use a new version. Upcoming FTE on MongoDB will upgrade the minimum supported version to 4.2.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 6]
```diff
@@ -151,6 +159,8 @@ public MongoSession(TypeManager typeManager, MongoClient client, MongoClientConf
         this.client = requireNonNull(client, "client is null");
         this.schemaCollection = requireNonNull(config.getSchemaCollection(), "config.getSchemaCollection() is null");
         this.caseInsensitiveNameMatching = config.isCaseInsensitiveNameMatching();
+        this.authorizedDatabases = config.isMetadataPrivilegesRequired() == false;
```
Comment (skrzypo987) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:162]: `!config.isMetadataPrivilegesRequired()`. Same below

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 7]
```diff
@@ -173,7 +183,8 @@ public List<HostAddress> getAddresses()
 
     public List<String> getAllSchemas()
     {
-        return ImmutableList.copyOf(client.listDatabaseNames()).stream()
+        return ImmutableList.copyOf(listDatabaseNames())
```
Comment (skrzypo987) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:186]: Rollback. `stream` in the same line is typical in Trino.
And BTW, always do those kind of changes in a different commit.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+    @AfterClass(alwaysRun = true)
+    public void tearDown()
+            throws Exception
+    {
+        try (afterClassCloser) {
+            //
+        }
+        finally {
+            queryRunnerWithDefaultSettings = null;
```
Comment (skrzypo987) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java:78]: Either `AutoClosable` or `finally` clause.
Move those 2 statements to setup method as a:
``` 
afterClassCloser.register(()->{
    queryRunnerWithDefaultSettings = null;
    queryRunner = null;
});
```
and just run `afterClassCloser.close()` here

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoClientConfig.java [hunk 9]
```diff
@@ -35,6 +35,7 @@ public void testDefaults()
                 .setCaseInsensitiveNameMatching(false)
                 .setSeeds("")
                 .setCredentials("")
+                .setMetadataPrivilegesRequired(true)
```
Comment (skrzypo987) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoClientConfig.java:38]: You should also put that add that property to `testExplicitPropertyMappings` method

**Review by skrzypo987 — COMMENTED** [review]
Looks ok, ping me for another round.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 10]
```diff
@@ -823,34 +834,62 @@ private String toRemoteSchemaName(String schemaName)
...
         }
         return schemaName;
     }
 
+    private Iterable<String> listDatabaseNames()
+    {
+        OperationExecutor operationExecutor = ((MongoClientImpl) client).getOperationExecutor();
+        CodecRegistry codecRegistry = ((MongoClientImpl) client).getSettings().getCodecRegistry();
+        boolean retryReads = ((MongoClientImpl) client).getSettings().getRetryReads();
+        return new ListDatabasesIterableImpl<>(null, BsonDocument.class, codecRegistry, ReadPreference.primary(), operationExecutor, retryReads)
```
Comment (Praveen2112) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:850]: Any reason on why are we extracting this here ?

Comment (Praveen2112) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:850]: Can we use `listDatabases` and apply the changes on top of them ?

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java (hunk 8 above)
Comment (Praveen2112) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java:78]: Can we make `TestMetadataPrivilegesRequired` to extend  `AbstractTestQueryFramework` ?

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 11]
```diff
@@ -823,34 +834,62 @@ private String toRemoteSchemaName(String schemaName)
...
             if (tableName.equals(remoteTableName.toLowerCase(ENGLISH))) {
                 return remoteTableName;
             }
         }
         return tableName;
     }
 
+    private Iterable<String> listCollectionNames(String databaseName)
+    {
+        Document listCollectionsCommand = new Document(ImmutableMap.<String, Object>builder()
```
Comment (Praveen2112) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:872]: Can we make this command as a static final variable ?

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java [hunk 12]
```diff
@@ -42,6 +50,11 @@ public ConnectionString getConnectionString()
         return new ConnectionString(dockerContainer.getReplicaSetUrl());
     }
 
+    public ConnectionString getConnectionString(String user, String password)
```
Comment (Praveen2112) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java:53]: Can we create a client directly instead of creating `ConnectionString` and creating based on it.

Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java:53]: It's also used to provide value for `mongodb.connection-url` property.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 10 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:850]: Sorry I do not understand what you mean. Could you explain more?

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java (hunk 12 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java:53]: I see `mongodb.credentials` is marked as deprecated and in my case using `mongodb.credentials` does not work in my test case.

Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java:53]: Both `mongodb.credentials` and `mongodb.connection-url` are annotated with `@ConfigSecuritySensitive`

Comment (Praveen2112) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java:53]: Makes sense!!

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 10 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:850]: Oh, I did not notice this before, I will change it

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 11 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:872]: Because `Document` is mutable I extracted ImmutableMap

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 2 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java]: I do not need this anymore

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java (hunk 4 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/MongoServer.java]: Renamed

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java (hunk 5 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java]: I set version to `4.4.17` - it seams docker init flakiness (when I set root user and password via env) got fixed

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 6 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java]: I removed property so this does not exists anymore

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoClientConfig.java (hunk 9 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoClientConfig.java]: I removed property so this does not exists anymore

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 10 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java]: Updated

**Comment by huberty89** [comment]
@skrzypo987 @ebyhr @Praveen2112 I make an update, please take a look

**Comment by ebyhr** [comment]
> Looks like TestMetadataPrivilegesRequired test added by me is flaky - Docker container with MongoDB cannot starts successfully.

How did you fix it? I'm facing the same issue.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java (hunk 5 above)
Comment (ebyhr) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java]: Hmm, it's still failing in my environment.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        }
+        catch (Throwable e) {
+            closeAllSuppress(e, queryRunner);
+            throw e;
+        }
+    }
+
+    private static void createTestRole(MongoDatabase database)
+    {
+        Double commandStatus = database.runCommand(new BsonDocument("createRole", new BsonString(TEST_ROLE))
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java:113]: Is it possible to use `Document` class to improve readability?

**Review by ebyhr — COMMENTED** [review]
Looks good except for the flaky container initialization & a minor comment.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 14]
```diff
@@ -837,34 +842,54 @@ private String toRemoteSchemaName(String schemaName)
...
-        for (String remoteSchemaName : client.listDatabaseNames()) {
+        for (String remoteSchemaName : listDatabaseNames()) {
             if (schemaName.equals(remoteSchemaName.toLowerCase(ENGLISH))) {
                 return remoteSchemaName;
             }
         }
         return schemaName;
     }
 
+    private Iterable<String> listDatabaseNames()
```
Comment (s2lomon) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:853]: Maybe instead returning `Iterable` from here and later turning it to Stream through `ImmutableList.of(listDatabaseNames()).stream()`, please return `Iterable` from here, for example by `Streams.stream(...)`.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 15]
```diff
@@ -178,7 +183,7 @@ public List<HostAddress> getAddresses()
 
     public List<String> getAllSchemas()
     {
-        return ImmutableList.copyOf(client.listDatabaseNames()).stream()
+        return ImmutableList.copyOf(listDatabaseNames()).stream()
```
Comment (s2lomon) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:186]: please look below  - change the return type of listDatabaseNames to make it simpler to read. I'm kind of lost on what is transformed into what.

**Review by s2lomon — COMMENTED** [review]
Please elaborate on why these changes are required in commit message. Right now it doesn't seem important and it is. In addition please describe what is the current  behaviour etc.

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java [hunk 16]
```diff
@@ -837,34 +842,54 @@ private String toRemoteSchemaName(String schemaName)
...
         }
         return tableName;
     }
 
+    private List<String> listCollectionNames(String databaseName)
+    {
+        MongoDatabase database = client.getDatabase(databaseName);
+        Document cursor = database.runCommand(new Document(AUTHORIZED_LIST_COLLECTIONS_COMMAND)).get("cursor", Document.class);
+
+        List<Document> firstBatch = cursor.get("firstBatch", List.class);
```
Comment (Praveen2112) [inline plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:880]: Would there be another batch ?

Comment (Praveen2112) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:880]: Or a next batch ?

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+                .getOnlyColumnAsSet())
+                .containsExactly("information_schema", TEST_DATABASE);
+    }
+
+    @Test
+    public void testTablesVisibility()
+    {
+        assertThat(getQueryRunner().execute("SHOW TABLES FROM mongodb." + TEST_DATABASE)
+                .getOnlyColumnAsSet())
+                .containsExactly(TEST_COLLECTION.toLowerCase(Locale.ENGLISH));
```
Comment (Praveen2112) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java:68]: Do we need a redundant conversion ? Can we use `assertQuery` where we could provide a result as `VALUES ...`

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+                    "mongodb.connection-url", connectionUrl));
+            return queryRunner;
+        }
+        catch (Throwable e) {
+            closeAllSuppress(e, queryRunner);
+            throw e;
+        }
+    }
+
+    private static void createTestRole(MongoDatabase database)
```
Comment (Praveen2112) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java:111]: Can we move this to `MongoServer`

**Review by skrzypo987 — COMMENTED** [review]
lgtm % minor comment and other's comments

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java (hunk 5 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMetadataPrivilegesRequired.java]: Looks like `org.testcontainers.containers.MongoDBContainer` contains a bug.

When I set `MONGO_INITDB_ROOT_USERNAME` and `MONGO_INITDB_ROOT_PASSWORD` this cause a restart MongoDB process.
At the same time `db.runCommand( { isMaster: 1 } ).ismaster==false` is executed inside Docker container in a loop (60 times) and during process restart this throw network error - that's the main issue.

I create `AuthenticatedMongoServer` which use `GenericContainer` directly. Default enabled `replication` is not needed in current scenario.

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java (hunk 13 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java]: Got it, I also try to improve readability by spiting it into few methods

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 14 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java]: We discuss this offline, because `listDatabaseNames` is private I return now `MongoIterable`

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java (hunk 18 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java]: I moved into `AuthenticatedMongoServer`

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java [hunk 19]
```diff
@@ -0,0 +1,107 @@
...
+            throws Exception
+    {
+        AuthenticatedMongoServer mongoServer = closeAfterClass(setupMongoServer());
+        return createMongoQueryRunner(mongoServer.testUserConnectionString().getConnectionString());
+    }
+
+    @Test
+    public void testSchemasVisibility()
+    {
+        assertThat(getQueryRunner().execute("SHOW SCHEMAS FROM mongodb")
```
Comment (ebyhr) [inline plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java:51]: This test succeeds even before the change. Is it possible to make the test that failed previously?

File: plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java (hunk 16 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:879]: I checked in a MongoDB Java driver source and by just creating 10,000 collections in MongoDB and by default MongoDB returns everything in a single response, so this implementation should be sufficient.

Comment (Praveen2112) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:879]: Does MongoDB takes too long to create 10,000 collections ? Can we add it as a test to ensure that if the newer version returns in some batch we could handle the same ?

Comment (huberty89) [reply plugin/trino-mongodb/src/main/java/io/trino/plugin/mongodb/MongoSession.java:879]: Listing 10,000 collections took for me 1.14 second. Still agree a test for this case is a good idea, I will add it

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java (hunk 19 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java:51]: Unfortunately I'm not able to configure MongoDB access control in such way to simulate case from here: https://www.mongodb.com/docs/manual/reference/command/listDatabases/#behavior for `MongoDB 4.0.6+`. I use `4.2.0`
I would still leave this test but as separate commit - I will extract this

File: plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java (hunk 17 above)
Comment (huberty89) [reply plugin/trino-mongodb/src/test/java/io/trino/plugin/mongodb/TestMongoPrivileges.java]: Replaced

**Comment by huberty89** [comment]
@Praveen2112 @skrzypo987 @ebyhr Please take a look if everything looks good and it's good to be merged, thanks

**Review by ebyhr — APPROVED** [review]

**Comment by ebyhr** [comment]
Merged, thanks!

**Comment by findepi** [comment]
Follow up: https://github.com/trinodb/trino/pull/15652
